### PR TITLE
Reference types

### LINKED ISSUE #100: Add support for reference types
This includes these types: 
```
BINARYEN_API BinaryenType BinaryenTypeFuncref(void);
BINARYEN_API BinaryenType BinaryenTypeExternref(void);
```

These ops:
```
BINARYEN_API BinaryenOp BinaryenRefIsNull(void);
BINARYEN_API BinaryenOp BinaryenRefIsFunc(void);
BINARYEN_API BinaryenOp BinaryenRefIsData(void);
BINARYEN_API BinaryenOp BinaryenRefIsI31(void);
BINARYEN_API BinaryenOp BinaryenRefAsNonNull(void);
BINARYEN_API BinaryenOp BinaryenRefAsFunc(void);
BINARYEN_API BinaryenOp BinaryenRefAsData(void);
BINARYEN_API BinaryenOp BinaryenRefAsI31(void);
```

And these expressions:
```
BINARYEN_API BinaryenExpressionRef BinaryenRefNull(BinaryenModuleRef module,
                                                   BinaryenType type);
BINARYEN_API BinaryenExpressionRef BinaryenRefIs(BinaryenModuleRef module,
                                                 BinaryenOp op,
                                                 BinaryenExpressionRef value);
BINARYEN_API BinaryenExpressionRef BinaryenRefAs(BinaryenModuleRef module,
                                                 BinaryenOp op,
                                                 BinaryenExpressionRef value);
BINARYEN_API BinaryenExpressionRef BinaryenRefFunc(BinaryenModuleRef module,
                                                   const char* func,
                                                   BinaryenType type);
BINARYEN_API BinaryenExpressionRef BinaryenRefEq(BinaryenModuleRef module,
                                                 BinaryenExpressionRef left,
                                                 BinaryenExpressionRef right);
```

### PR DESCRIPTION
**Pull request by jozanza**
Getting this PR going. Adding support for reference types as described in #100 

Closes #100

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,5 +1,6 @@
 node_modules/
 _esy/
+_build/
```
Comment (phated) [inline .gitignore:3]: opam? 😛

Comment (jozanza) [reply .gitignore:3]: I have no idea what generated the `_build/` dir tbh, but that sounds right to me 😂 lol

**Comment by phated** [comment]
@jozanza @ospencer is there a reason this is still a draft? Just waiting for the other stuff to land?

**Comment by jozanza** [comment]
@phated No reason! I just changed it to "Ready for Review" 👍🏽

**Review by ospencer — APPROVED** [review]

**Review by phated — APPROVED** [review]
I needed to rebase on master and make a fix to the tests, but LGTM now!

**Review by ospencer — APPROVED** [review]
🎉

**Pull request merged**